### PR TITLE
[WIP]: IPv4 matcher

### LINKED ISSUE #2722: Ready-made matchers for IP addresses
A recurring pain point is writing matchers for IPv4, IPv6, email addresses, etc. This leads to cargo culting of regexps, often wrong, from stackexchange. Even worse is trying to find valid regex for CIDR matching.

https://mobile.twitter.com/bettersafetynet/status/1312884924811227137 reminded me that we have the chance to get this right for our users days one, so we should :)

### PR DESCRIPTION
**Pull request by kavirajk**
**What this PR does / why we need it**:
Support built-in IP matcher on LogQL

**Which issue(s) this PR fixes**:
Fixes #2722 

**Special notes for your reviewer**:
This PR support only IPV4 matchers. IPV6 will be done in separate PR

**Checklist**
- [ ] Make it work
- [ ] Make it clean
- [ ] Make it fast
- [ ] Documentation added
- [ ] Tests updated

### CODE REVIEW

File: pkg/logql/ast.go [hunk 1]
```diff
@@ -226,6 +226,56 @@ func (e *pipelineExpr) HasFilter() bool {
...
+	ty     labels.MatchType
+	op     string
+	params string
+	implicit
+}
+
+func newFunctionLineFilterExpr(op, params string, ty labels.MatchType) *functionLineFilterExpr {
+	switch op {
+	case OpFilterIP:
+		// todo filter panic(newParseError("this is not an ip or range", 0, 0))
```
Comment (cyriltovena) [inline pkg/logql/ast.go:239]: something's missing here.

File: pkg/logql/log/ip.go [hunk 2]
```diff
@@ -0,0 +1,195 @@
...
+func NewLineIpFilter(pattern string) (Stage, error) {
+	return &LineIPFiler{pat: pattern}, nil
+}
+
+func (f *LineIPFiler) Process(line []byte, _ *LabelsBuilder) ([]byte, bool) {
+	match, _ := ip(f.pat, string(line))
+	if match {
+		return line, true
+	}
+	return line, false
```
Comment (cyriltovena) [inline pkg/logql/log/ip.go:24]:
```suggestion
return line, match
```

File: pkg/logql/log/ip.go [hunk 3]
```diff
@@ -0,0 +1,195 @@
...
+// This actually implements the Stage interface to process the line
+type LineIPFiler struct {
+	pat string
+}
+
+func NewLineIpFilter(pattern string) (Stage, error) {
+	return &LineIPFiler{pat: pattern}, nil
+}
+
+func (f *LineIPFiler) Process(line []byte, _ *LabelsBuilder) ([]byte, bool) {
```
Comment (cyriltovena) [inline pkg/logql/log/ip.go:19]: I'd like to see a benchmark of this. I wonder if the `string(line)` creates a copy or not. If so I think you should avoid it.

File: pkg/logql/log/ip.go [hunk 4]
```diff
@@ -0,0 +1,195 @@
...
+	match, _ := ip(f.pat, string(line))
```
Comment (cyriltovena) [inline pkg/logql/log/ip.go:20]: The parsing of the AST parameter (e.g ip("127.0.0.1")) should not be for every line but instead only once.

This should also be returned as an AST error if the format is invalid.

File: pkg/logql/log/ip.go [hunk 5]
```diff
@@ -0,0 +1,195 @@
...
+import (
+	"errors"
+	"fmt"
+	"strings"
+	"unicode"
+)
+
+// This actually implements the Stage interface to process the line
+type LineIPFiler struct {
+	pat string
```
Comment (cyriltovena) [inline pkg/logql/log/ip.go:12]: I don't even think you need to keep path but instead the parsed version.

File: pkg/logql/log/ip.go [hunk 6]
```diff
@@ -62,30 +64,54 @@ func ip(pat, line string) (bool, error) {
...
+		ip [4]uint
+		i  int
+	)
+	p.buf.Reset()
+	for _, r := range s {
+		if p.buf.Len() > 4 {
+			return ip, errInValidIP
+		}
+		if r == '.' { // 127.0.0.1
+			strconv.ParseInt(unsafeGetString(p.buf.Bytes()), 10, 32)
```
Comment (sandeepsukhani) [inline pkg/logql/log/ip.go:106]: Am I reading this wrong or did you miss assigning the value of this to `ip`?

Comment (kavirajk) [reply pkg/logql/log/ip.go:106]: you are right. should be assigning it to `ip[i]`. thanks for the catch :)

Comment (cyriltovena) [reply pkg/logql/log/ip.go:106]: 🤭

**Comment by oivindoh** [comment]
This is 🚀 👍 - wouldn't it make sense to integrate ipv6 in the same PR?

**Comment by kavirajk** [comment]
Close this in favour of #3986

**Pull request closed without merging**